### PR TITLE
Improving PCL_DEPRECATED with encorcable deprecation policy

### LINKED ISSUE #3805: Better deprecation policy
**Is your feature request related to a problem? Please describe.**

It's hard to know when something got deprecated or when it would be removed (just from reading the code)

**Context**

If we can add the information regarding the cycle in which deprecation was added, we can automate removal of deprecated items too.

**Describe the solution you'd like**

[Working example](https://godbolt.org/z/U9Dyt3)
```cpp
#define DEPRECATED_IMPL_YES(msg) [[deprecated(msg)]]
#define DEPRECATED_IMPL_NO(msg) deprecation_over("Remove the deprecated content") \
DEPRECATED_IMPL_YES(msg)

#define DEPRECATED(version, msg) DEPRECATED_##version(msg)

#define DEPRECATED_1_10(msg) DEPRECATED_IMPL_YES(msg)
#define DEPRECATED_1_9(msg) DEPRECATED_IMPL_NO(msg)

DEPRECATED(1_10, "msg")
```
The `DEPRECATED_X_Y` can be generated using Boost.PP since they depend on the current minor version. For current version `1_10`, we should be removing all items from `1_8` (2 cycles before, this can be codified using a macro constant). This will force the release PR to remove the deprecation when the version is bumped to `1_11` or `2_0` but not if there's a patch release (eg: `1.10.1`).

`DEPRECATED_IMPL_YES` is the current macro `PCL_DEPRECATED`
`DEPRECATED_IMPL_NO` causes a simple compile error

### PR DESCRIPTION
**Pull request by kunaltyagi**
Closes #3805

### CODE REVIEW

**Comment by kunaltyagi** [comment]
This will fail to compile even. Put out in the open for comments

File: common/include/pcl/pcl_macros.h [hunk 1]
```diff
@@ -67,24 +67,53 @@
...
+                unneeded_deprecation)
+
+/**
+ * \brief macro for compatibility across compilers and help remove old deprecated
+ *        items before release based on PCL_DEPRECATION_CYCLE
+ *
+ * \detail compiler errors of `unneeded_deprecation` and `major_version_mismatch`
+ * are hints to the developer that those items ca be safely removed
+ */
+#define PCL_DEPRECATED(major, minor, message) \
```
Comment (taketwo) [inline common/include/pcl/pcl_macros.h]: `major` and `minor` are what exactly? Last version where the item was not yet deprecated? First version where it was deprecated? I'm not sure how we will choose this version given that we don't always know what the next released version will end up being (patch or minor).

I also feel like `message` is more important should be the first argument.

Finally, we may consider appending (computed) target removal version to the message.

File: common/include/pcl/pcl_macros.h [hunk 2]
```diff
@@ -67,24 +67,53 @@
...
-// PCL_DEPRECATED macro to [[deprecated]].
+// PCL_DEPRECATED_IMPL macro to [[deprecated]].
 #if defined(__has_cpp_attribute) && __has_cpp_attribute(deprecated) && !defined(__CUDACC__)
-  #define PCL_DEPRECATED(message) [[deprecated(message)]]
+  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
 #elif defined(__GNUC__) || defined(__clang__)
-  #define PCL_DEPRECATED(message) __attribute__((deprecated(message)))
+  #define PCL_DEPRECATED_IMPL(message) __attribute__((deprecated(message)))
 #elif defined(_MSC_VER)
   // Until Visual Studio 2013 you had to use __declspec(deprecated).
```
Comment (SunBlack) [inline common/include/pcl/pcl_macros.h]: I believe we can remove this comment. In general why not make it `#define PCL_DEPRECATED(message) [[deprecated(message)]]` for all compilers? We require C++14 and this is [supported](https://en.cppreference.com/w/cpp/compiler_support) by any relevant C++14 compiler.

File: common/include/pcl/pcl_macros.h (hunk 1 above)
Comment (kunaltyagi) [reply common/include/pcl/pcl_macros.h]: > major and minor are what exactly?

First version where it was deprecated, as per implementation. I can switch it to mean till when it'd be supported. Then I can convert the message to: "slated for removal in vX.Y. <usr msg>"

File: common/include/pcl/pcl_macros.h (hunk 2 above)
Comment (kunaltyagi) [reply common/include/pcl/pcl_macros.h]: `nvcc` is sadly still relevant

Comment (SunBlack) [reply common/include/pcl/pcl_macros.h]: Isn't it possible then to change the conditions so, that nvcc get's the old flags, but the other compilers are using the new attribute? Currently no compiler will use `[[deprecated(message)]]` as we force the compiler to allow only C++14 code, but `__has_cpp_attribute` is C++20.

File: common/include/pcl/pcl_macros.h (hunk 1 above)
Comment (taketwo) [reply common/include/pcl/pcl_macros.h]: Hm, I think I like this. No need to do arithmetics in macro and `PCL_DEPRECATION_CYCLE`.

File: common/include/pcl/pcl_macros.h (hunk 2 above)
Comment (kunaltyagi) [reply common/include/pcl/pcl_macros.h]: > but `__has_cpp_attribute` is C++20.

[No, it's C++14](https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B14)

Comment (SunBlack) [reply common/include/pcl/pcl_macros.h]: Ah sorry, just saw the headline [here](https://en.cppreference.com/w/cpp/feature_test).

File: common/include/pcl/pcl_macros.h (hunk 1 above)
Comment (kunaltyagi) [reply common/include/pcl/pcl_macros.h]: PTAL

File: common/include/pcl/pcl_macros.h [hunk 3]
```diff
@@ -67,24 +67,56 @@
...
-  #warning "You need to implement PCL_DEPRECATED for this compiler"
-  #define PCL_DEPRECATED(message)
+  #warning "You need to implement PCL_DEPRECATED_IMPL for this compiler"
+  #define PCL_DEPRECATED_IMPL(message)
 #endif
 
+#define PCL_DEPRECATED_MODIFY_MSG(major, minor, msg) \
+    "Slated for removal in PCL " \
+    BOOST_PP_STRINGIZE(PCL_MAJOR_VERSION.PCL_MINOR_VERSION) \
+    ": " msg
```
Comment (taketwo) [inline common/include/pcl/pcl_macros.h]: The warnings emitted by compilers (at least GCC) have the following structure:

> `LOCATION`: warning: '`SYMBOL`' is deprecated: `MSG` [-Wdeprecated-declarations]

Usually, our `msg` reads like "use xyz() instead", thus in total the emitted output becomes:

```
<source>:25:22: warning: 'int goo()' is deprecated: use xyz() instead [-Wdeprecated-declarations]
```

Are we sure we want to put this "Slated for removal..." message in front of `msg` and not after?

File: common/include/pcl/pcl_macros.h (hunk 2 above)
Comment (kunaltyagi) [reply common/include/pcl/pcl_macros.h]: Please unresolve this if your concerns weren't addressed. 😄

**Comment by taketwo** [comment]
LGTM. To my taste, I'd spell out "it will" completely, looks a bit more formal.

**Comment by kunaltyagi** [comment]
<details><summary>57 instances. We just need to determine when they need to get removed.</summary>

```
common/include/pcl/impl/point_types.hpp:1160:    PCL_DEPRECATED("Use ctor accepting all position (x, y, z) data")
common/include/pcl/impl/point_types.hpp:1164:    PCL_DEPRECATED("Use ctor accepting all viewpoint (vp_x, vp_y, vp_z) data")
common/include/pcl/point_cloud.h:140:    PCL_DEPRECATED("use createMapping() instead")
common/include/pcl/point_cloud.h:608:      PCL_DEPRECATED("rewrite your code to avoid using this protected field") shared_ptr<MsgFieldMap> mapping_;
common/include/pcl/types.h:52:  using uint8_t PCL_DEPRECATED("use std::uint8_t instead of pcl::uint8_t") = std::uint8_t;
common/include/pcl/types.h:53:  using int8_t PCL_DEPRECATED("use std::int8_t instead of pcl::int8_t") = std::int8_t;
common/include/pcl/types.h:54:  using uint16_t PCL_DEPRECATED("use std::uint16_t instead of pcl::uint16_t") = std::uint16_t;
common/include/pcl/types.h:55:  using int16_t PCL_DEPRECATED("use std::uint16_t instead of pcl::int16_t") = std::int16_t;
common/include/pcl/types.h:56:  using uint32_t PCL_DEPRECATED("use std::uint32_t instead of pcl::uint32_t") = std::uint32_t;
common/include/pcl/types.h:57:  using int32_t PCL_DEPRECATED("use std::int32_t instead of pcl::int32_t") = std::int32_t;
common/include/pcl/types.h:58:  using uint64_t PCL_DEPRECATED("use std::uint64_t instead of pcl::uint64_t") = std::uint64_t;
common/include/pcl/types.h:59:  using int64_t PCL_DEPRECATED("use std::int64_t instead of pcl::int64_t") = std::int64_t;
common/include/pcl/types.h:60:  using int_fast16_t PCL_DEPRECATED("use std::int_fast16_t instead of pcl::int_fast16_t") = std::int_fast16_t;
common/include/pcl/pcl_macros.h:108: * Warning message with PCL_DEPRECATED(1, 99, "Not needed anymore") till PCL 1.98:
common/include/pcl/pcl_macros.h:111:#define PCL_DEPRECATED(Major, Minor, Message)                       \
common/include/pcl/pcl_macros.h:154:PCL_DEPRECATED("use std::isnan instead of pcl_isnan")
common/include/pcl/pcl_macros.h:158:PCL_DEPRECATED("use std::isfinite instead of pcl_isfinite")
common/include/pcl/pcl_macros.h:162:PCL_DEPRECATED("use std::isinf instead of pcl_isinf")
common/include/pcl/common/io.h:79:  PCL_DEPRECATED("use getFieldIndex<PointT> (field_name, fields) instead")
common/include/pcl/common/io.h:109:  PCL_DEPRECATED("use getFields<PointT> () with return value instead")
common/include/pcl/common/io.h:119:  PCL_DEPRECATED("use getFields<PointT> () with return value instead")
common/include/pcl/common/io.h:331:  PCL_DEPRECATED("use pcl::concatenate() instead, but beware of subtle difference in behavior (see documentation)")
doc/doxygen/doxyfile.in:315:                         "PCL_DEPRECATED(x)=/** \deprecated x */"
features/include/pcl/features/rsd.h:93:  PCL_DEPRECATED("use computeRSD() overload that takes input point clouds by const reference")
features/include/pcl/features/rsd.h:119:  PCL_DEPRECATED("use computeRSD() overload that takes input point cloud by const reference")
filters/include/pcl/filters/passthrough.h:290:      PCL_DEPRECATED("use inherited FilterIndices::setNegative() instead")
filters/include/pcl/filters/passthrough.h:300:      PCL_DEPRECATED("use inherited FilterIndices::getNegative() instead")
filters/include/pcl/filters/passthrough.h:310:      PCL_DEPRECATED("use inherited FilterIndices::getNegative() instead")
io/include/pcl/io/hdl_grabber.h:74:      using sig_cb_velodyne_hdl_scan_point_cloud_xyzrgb PCL_DEPRECATED("use sig_cb_velodyne_hdl_scan_point_cloud_xyzrgba instead")
io/include/pcl/io/hdl_grabber.h:100:      using sig_cb_velodyne_hdl_sweep_point_cloud_xyzrgb PCL_DEPRECATED("use sig_cb_velodyne_hdl_sweep_point_cloud_xyzrgba instead")
io/include/pcl/io/ascii_io.h:120:      PCL_DEPRECATED("use parameterless setInputFields<PointT>() instead")
io/include/pcl/compression/entropy_range_coder.h:173:      PCL_DEPRECATED("use std::log2 instead")
octree/include/pcl/octree/octree_base.h:129:  PCL_DEPRECATED("use leaf_depth_begin() instead")
octree/include/pcl/octree/octree_base.h:136:  PCL_DEPRECATED("use leaf_depth_end() instead") const LeafNodeIterator leaf_end()
octree/include/pcl/octree/octree_base.h:692:  PCL_DEPRECATED("use std::log2 instead") double Log2(double n_arg)
octree/include/pcl/octree/octree2buf_base.h:240:  PCL_DEPRECATED("use leaf_depth_begin() instead")
octree/include/pcl/octree/octree2buf_base.h:247:  PCL_DEPRECATED("use leaf_depth_end() instead") const LeafNodeIterator leaf_end()
octree/include/pcl/octree/octree2buf_base.h:941:  PCL_DEPRECATED("use std::log2 instead") inline double Log2(double n_arg)
recognition/include/pcl/recognition/mask_map.h:81:  PCL_DEPRECATED("Use new version diff getDifferenceMask(mask0, mask1)")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:214:      PCL_DEPRECATED("remove PointNT from template parameters")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:223:      PCL_DEPRECATED("EuclideadClusterComparator never actually used normals and angular threshold, this function has no effect on the behavior of the comparator. It is deprecated and will be removed in future releases.")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:228:      PCL_DEPRECATED("EuclideadClusterComparator never actually used normals and angular threshold, this function has no effect on the behavior of the comparator. It is deprecated and will be removed in future releases.")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:235:      PCL_DEPRECATED("EuclideadClusterComparator never actually used normals and angular threshold, this function has no effect on the behavior of the comparator. It is deprecated and will be removed in future releases.")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:243:      PCL_DEPRECATED("EuclideadClusterComparator never actually used normals and angular threshold, this function has no effect on the behavior of the comparator. It is deprecated and will be removed in future releases.")
segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:250:      PCL_DEPRECATED("use setExcludeLabels(const ExcludeLabelSetConstPtr &) instead")
segmentation/include/pcl/segmentation/segment_differences.h:63:  PCL_DEPRECATED("tgt parameter is not used; it is deprecated and will be removed in future releases")
segmentation/include/pcl/segmentation/organized_multi_plane_segmentation.h:288:      PCL_DEPRECATED("centroids and covariances parameters are not used; they are deprecated and will be removed in future releases")
segmentation/include/pcl/segmentation/supervoxel_clustering.h:198:      PCL_DEPRECATED("constructor with flag for using the single camera transform is deprecated. Default behavior is now to use the transform for organized clouds, and not use it for unorganized. Use setUseSingleCameraTransform() to override the defaults.")
segmentation/include/pcl/segmentation/supervoxel_clustering.h:282:      PCL_DEPRECATED("use getLabeledCloud() instead. An example of how to display and save with colorized labels can be found in examples/segmentation/example_supervoxels.cpp")
surface/include/pcl/surface/mls.h:360:      PCL_DEPRECATED("use setPolynomialOrder() instead")
surface/include/pcl/surface/mls.h:378:      PCL_DEPRECATED("use getPolynomialOrder() instead")
surface/include/pcl/surface/mls.h:750:  using MovingLeastSquaresOMP PCL_DEPRECATED("use MovingLeastSquares instead, it supports OpenMP now") = MovingLeastSquares<PointInT, PointOutT>;
surface/src/3rdparty/opennurbs/opennurbs_annotation2.cpp:5825:PCL_DEPRECATED("Use the version that takes a model transform argument")
surface/src/3rdparty/opennurbs/opennurbs_annotation2.cpp:5865:PCL_DEPRECATED("Use the version that takes a dimstyle pointer")
surface/src/3rdparty/opennurbs/opennurbs_annotation2.cpp:5994:PCL_DEPRECATED("Use the version that takes a model transform argument")
visualization/include/pcl/visualization/point_cloud_color_handlers.h:117:        PCL_DEPRECATED("use getColor() without parameters instead")
visualization/include/pcl/visualization/point_cloud_color_handlers.h:619:        PCL_DEPRECATED("use getColor() without parameters instead")
```
</details>

**Comment by SunBlack** [comment]
As there was no announcement until now I suggest to all 1.12 or 1.13. So two minor versions from now.

**Comment by SergioRAgostinho** [comment]
The way I usually went about things was to grep the lib for PCL_DEPRECATED, jump to github, see the commit history and check which first version it got included in.

Example:
This [one](https://github.com/PointCloudLibrary/pcl/blame/4875a4819ce7a23a7e7136ec7b6adda18ea3a253/common/include/pcl/point_cloud.h#L141) got added for 1.10 https://github.com/PointCloudLibrary/pcl/commit/5f4561e0bcf73321ac83214c7015129714373790

**Comment by taketwo** [comment]
> As there was no announcement until now I suggest to all 1.12 or 1.13. So two minor versions from now.

Starting from 1.9.0, we include the following sentence in the changelog:

> Deprecated code scheduled to be removed after two minor releases.

It's open for interpretation though what precisely "after two minor releases" means, `+2` or `+3` to the minor number.

**Comment by kunaltyagi** [comment]
> It's open for interpretation though what precisely "after two minor releases" means

Let's go for 1.12 in that case.

**Comment by SunBlack** [comment]
I saw this message now during compiling:
```
In file included from /home/sunblack/dev/pcl/tools/mesh_sampling.cpp:38:
/home/sunblack/dev/pcl/visualization/include/pcl/visualization/pcl_visualizer.h:74:31: warning: 'mapping_' is deprecated: rewrite your code to avoid using this protected field (It will be removed in PCL 1.10) [-Wdeprecated-declarations]
  template <typename T> class PointCloud;
                              ^
/home/sunblack/dev/pcl/common/include/pcl/range_image/range_image.h:55:9: note: in implicit copy constructor for 'pcl::PointCloud<pcl::PointWithRange>' first required here
  class RangeImage : public pcl::PointCloud<PointWithRange>
        ^
/home/sunblack/dev/pcl/common/include/pcl/range_image/range_image.h:125:39: note: in implicit copy constructor for 'pcl::RangeImage' first required here
      makeShared () { return Ptr (new RangeImage (*this)); }
                                      ^
/home/sunblack/dev/pcl/common/include/pcl/point_cloud.h:608:7: note: 'mapping_' has been explicitly marked deprecated here
      PCL_DEPRECATED(1, 12, "rewrite your code to avoid using this protected field") shared_ptr<MsgFieldMap> mapping_;
      ^
/home/sunblack/dev/pcl/common/include/pcl/pcl_macros.h:122:17: note: expanded from macro 'PCL_DEPRECATED'
                PCL_DEPRECATED_MINOR(Minor,                             \
                ^
/home/sunblack/dev/pcl/common/include/pcl/pcl_macros.h:108:17: note: expanded from macro 'PCL_DEPRECATED_MINOR'
                PCL_DEPRECATED_IMPL(Msg),           \
                ^
/home/sunblack/dev/pcl/common/include/pcl/pcl_macros.h:89:42: note: expanded from macro 'PCL_DEPRECATED_IMPL'
  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
                                         ^
```

Relevant parts: `'mapping_' is deprecated: rewrite your code to avoid using this protected field (It will be removed in PCL 1.10)` vs `PCL_DEPRECATED(1, 12, "rewrite your code to avoid using this protected field") `. Any idea what happened here? Compiler is Clang 10 ;-)

**Comment by aPonza** [comment]
The deprecation list is showing weird things on https://pointcloudlibrary.github.io/documentation/deprecated.html but not on http://docs.pointclouds.org/trunk/deprecated.html.

**Comment by kunaltyagi** [comment]
@SunBlack create a fix PR. It was a tragedy of errors of my part.

@aPonza I'll have to take a deeper look at that. Might take some time

**Comment by kunaltyagi** [comment]
> doxygen could be confused by a macro call without semicolon

Wasn't this error being reported before?

Based on the 2 deprecated lists you linked, I think somewhere doxygen is missing a ')' and is picking one from the deprecated message itself.

**Comment by aPonza** [comment]
It used to be there before the part1 documentation fix pr which removed all of them iirc.

**Comment by kunaltyagi** [comment]
But that [PR](https://github.com/PointCloudLibrary/pcl/pull/3701/files?diff=split&w=1) #3701 didn't touch the same lines I did (in the header file).

**Comment by divmadan** [comment]
A query. Is this line correct?
https://github.com/PointCloudLibrary/pcl/blob/146f95c7109e6b8d3f452d0d26695923eb5b95f5/common/include/pcl/pcl_macros.h#L118

**Comment by kunaltyagi** [comment]
#3945 addresses this (and more)

**Pull request merged**